R/twas.R

```diff
@@ -353,46 +351,70 @@ harmonize_twas <- function(twas_weights_data, ld_meta_file_path, gwas_meta_file,
-#' A dataframe of twas results summary is generated for each gene-contexts-method pair of all methods for imputable genes.
-#' input returned.
 #' @param twas_weights_data List of list of twas weights output from generate_twas_db function.
 #' @param region_block A string with informaiton of chromosome number, startind position, and ending position of LD block conneced with "_".
-#' @param twas_data_loader A data loader class object that load twas weights data for the expected format to be qc/harmonized.
+#' @return A list of list containing twas result table and formatted input data for ctwas_sumstats main function.
+#' \itemize{
+#'   \item{twas_table}{ A dataframe of twas results summary is generated for each gene-contexts-method pair of all methods for imputable genes.}
+#'   \item{ctwas_db}{ A list of list containing pre-processed input data for ctwas_sumstats function. }
+#' }
 #' @export
 twas_pipeline <- function(twas_weights_data,
                           ld_meta_file_path,
                           gwas_meta_file,
-                          twas_weights_loader,
                           region_block) {
   # Step 1: TWAS analysis for all methods for imputable gene
-  twas_result_table <- do.call(rbind, lapply(names(twas_weights_data), function(weight_db) {
+  twas_results_db <- lapply(names(twas_weights_data), function(weight_db) {
     # harmonize twas weights and gwas sumstats against LD
-    twas_data_qced <- harmonize_twas(twas_weights_data[[weight_db]], ld_meta_file_path, gwas_meta_file,
-      twas_data_loader = twas_weights_loader
-    )
+    twas_data_qced <- harmonize_twas(twas_weights_data[[weight_db]], ld_meta_file_path, gwas_meta_file)
     gene <- names(twas_data_qced)
     if (length(gene) < 1) stop(paste0("No gene's data processed at harmonization process for ", weight_db, ". "))
     contexts <- names(twas_data_qced[[gene]][["weights_qced"]])
     twas_gene_table <- do.call(rbind, lapply(contexts, function(context) {
       twas_contexts <- do.call(rbind, lapply(gwas_studies, function(study) {
+        # twas analysis
         twas_rs <- twas_analysis(
           twas_data_qced[[gene]][["weights_qced"]][[context]][["weights"]], twas_data_qced[[gene]][["gwas_qced"]][[study]],
           twas_data_qced[[gene]][["LD"]], rownames(twas_data_qced[[gene]][["weights_qced"]][[context]][["weights"]])
         )
+        # summarize twas results by methods for a context within a gene
         context_table <- data.frame(
           gwas_study = study, method = sub("_[^_]+$", "", names(twas_rs)), twas_z = find_data(twas_rs, c(2, "z")),
           twas_pval = find_data(twas_rs, c(2, "pval"))
         )
         return(context_table)
       }))
+      # mr analysis - for a single context
+      mr_formatted_input <- mr_format(twas_weights_results[[gene]]$susie_results, context, twas_data_qced[[gene]][["gwas_qced"]][[study]],
```

Comment:
For MR, I think the idea is to only do it for those that are TWAS significant.

Reply:
You can filter for that with some user specified TWAS p-value cutoff and only perform MR beyond that cutoff.

Reply:
and set default p-value cutoff for mr analysis be 0.05?